### PR TITLE
[PyTorch][Training][EC2] PyTorch 2.5.1 Currency Release

### PR DESCRIPTION
**Pull request by sirutBuasai**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
DLC features the following package versions:
- Python 3.11.10
- PyTorch 2.5.1
- Miniforge3 24.9.0
- CUDA 12.4.1
- cuDNN 9.1.0.70
- NCCL 2.23.4
- EFA 1.36.0
- hwloc 2.11.2
- aws-ofi-nccl 1.12.1
- GDRCopy 2.4.2
- TransformerEngine 1.11
- Flash Attention 2.6.3

### Tests run
- [486ca40](https://github.com/aws/deep-learning-containers/pull/4413/commits/486ca40108e2e58b81c2804d8992d2cf0b58cee5) Passed benchmark, sanity, security, eks, ecs tests
- [c6ce319](https://github.com/aws/deep-learning-containers/pull/4413/commits/c6ce3190844535547e0cd265414c146d4a1f7c16) Passed ec2 test

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

<details>
<summary>Confused on how to run tests? Try using the helper utility...</summary>

Assuming your remote is called `origin` (you can find out more with `git remote -v`)...
  
- Run default builds and tests for a particular buildspec - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -cp origin`

- Enable specific tests for a buildspec or set of buildspecs - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -t sanity_tests -cp origin`

- Restore TOML file when ready to merge

`python src/prepare_dlc_dev_environment.py -rcp origin`
</details>

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Fill out the template and click the checkbox of the builds you'd like to execute

*Note: Replace with <X.Y> with the major.minor framework version (i.e. 2.2) you would like to start.*

- [ ] build_pytorch_training_<X.Y>_sm
- [ ] build_pytorch_training_<X.Y>_ec2

- [ ] build_pytorch_inference_<X.Y>_sm
- [ ] build_pytorch_inference_<X.Y>_ec2
- [ ] build_pytorch_inference_<X.Y>_graviton

- [ ] build_tensorflow_training_<X.Y>_sm
- [ ] build_tensorflow_training_<X.Y>_ec2

- [ ] build_tensorflow_inference_<X.Y>_sm
- [ ] build_tensorflow_inference_<X.Y>_ec2
- [ ] build_tensorflow_inference_<X.Y>_graviton
</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by Lokiiiiii — APPROVED** [review]

File: test/dlc_tests/container_tests/bin/pytorch_tests/testNVApex [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 
+set -e
+
 HOME_DIR=/test
 BIN_DIR=${HOME_DIR}/bin
 
 echo "Testing Nvidia Apex imports"
 
-python -c "import torch; import apex; from apex import amp; from apex.parallel import DistributedDataParallel; from apex import optimizers; from apex.fp16_utils import *"
+python -c "import torch; import apex; from apex import amp; from apex.parallel import DistributedDataParallel; from apex import optimizers; from apex.fp16_utils import *" || exit 1
```
Comment (roywei) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testNVApex:10]: shall we just disable this test properly if we are removing apex?

Comment (sirutBuasai) [reply test/dlc_tests/container_tests/bin/pytorch_tests/testNVApex:10]: It's disabled.

File: pytorch/training/docker/2.5/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    "setuptools>=70.0.0" \
+    "urllib3<2" \
+    "awscli<2" \
+ && /opt/conda/bin/mamba clean -afy \
+ && rm -rf /etc/apt/sources.list.d/*
+
+# Install common pip packages (in case of conda package is not available)
+RUN pip install --no-cache-dir opencv-python mpi4py
+
+RUN curl -o /license.txt https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.4/license.txt
```
[the remark this answers is not in the document]
Comment (Yadan-Wei) [reply pytorch/training/docker/2.5/py3/Dockerfile.cpu:179]: Will update with a new PR.

File: pytorch/training/docker/2.5/py3/cu124/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+    "awscli<2" \
+    libgcc \
+    ninja \
+ && /opt/conda/bin/mamba clean -afy \
+ && rm -rf /etc/apt/sources.list.d/*
+
+# Install common pip packages (in case of conda package is not available)
+RUN pip install --no-cache-dir opencv-python mpi4py
+
+RUN curl -o /license.txt https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.4/license.txt
```
[the remark this answers is not in the document]
Comment (Yadan-Wei) [reply pytorch/training/docker/2.5/py3/cu124/Dockerfile.gpu:260]: Update with a new PR.

**Review by Yadan-Wei — APPROVED** [review]

**Pull request merged**